File: test/Cartesian3d.cpp

#include <catch2/catch.hpp>

#define EQ(a,b) REQUIRE_THAT((a), Catch::WithinAbs((b), 1.e-12));

#include <GMatElastoPlasticFiniteStrainSimo/Cartesian3d.h>

namespace GM = GMatElastoPlasticFiniteStrainSimo::Cartesian3d;


TEST_CASE("GMatElastoPlasticFiniteStrainSimo::Cartesian3d", "Cartesian3d.h")
{

double K = 12.3;
double G = 45.6;

double gamma = 0.02;

GM::Tensor2 F;
F.fill(0.0);
F(0,0) = 1.0 + gamma;
F(1,1) = 1.0 / (1.0 + gamma);
F(2,2) = 1.0;


SECTION("Elastic")
{
    GM::Elastic mat(K, G);

    auto Sig = mat.Stress(F);

    EQ(Sig(0,0), G * +2.0 * std::log(1.0 + gamma));
    EQ(Sig(1,1), G * -2.0 * std::log(1.0 + gamma));
    EQ(Sig(2,2), 0.0);
    EQ(Sig(0,1), 0.0);
    EQ(Sig(0,2), 0.0);
    EQ(Sig(1,0), 0.0);
    EQ(Sig(1,2), 0.0);
    EQ(Sig(2,0), 0.0);
    EQ(Sig(2,1), 0.0);
}


SECTION("Matrix")
{
    size_t nelem = 3;
    size_t nip = 2;

    GM::Matrix mat(nelem, nip);

    // all rows elastic
    {
        xt::xtensor<size_t,2> I = xt::ones<size_t>({nelem, nip});
        mat.setElastic(I, K, G);
    }

    xt::xtensor<double,4> f = xt::empty<double>({nelem, nip, 3ul, 3ul});

    for (size_t e = 0; e < nelem; ++e) {
        for (size_t q = 0; q < nip; ++q) {
            xt::view(f, e, q) = F;
        }
    }

    auto sig = mat.Stress(f);

    for (size_t e = 0; e < nelem; ++e) {
        for (size_t q = 0; q < nip; ++q) {
            EQ(sig(e,q,0,0), G * +2.0 * std::log(1.0 + gamma));
            EQ(sig(e,q,1,1), G * -2.0 * std::log(1.0 + gamma));
            EQ(sig(e,q,0,1), 0.0);
            EQ(sig(e,q,1,0), 0.0);
        }
    }

    REQUIRE(xt::allclose(xt::view(sig, xt::all(), xt::all(), 2, xt::all()), 0.0));
    REQUIRE(xt::allclose(xt::view(sig, xt::all(), xt::all(), xt::all(), 2), 0.0));
}


}

File: test/Cartesian3d.py

import GMatElastoPlasticFiniteStrainSimo.Cartesian3d as GMat
import numpy as np


def EQ(a,b):
    assert np.abs(a-b) < 1.e-12

def ALLEQ(a, b):
    assert np.allclose(a, b)


K = 12.3
G = 45.6

gamma = 0.02

F = np.array([
        [1.0 + gamma, 0.0, 0.0],
        [0.0, 1.0 / (1.0 + gamma), 0.0],
        [0.0, 0.0, 1.0]])


# Elastic


mat = GMat.Elastic(K, G)

Sig = mat.Stress(F)

EQ(Sig[0,0], G * +2.0 * np.log(1.0 + gamma))
EQ(Sig[1,1], G * -2.0 * np.log(1.0 + gamma))
EQ(Sig[2,2], 0.0)
EQ(Sig[0,1], 0.0)
EQ(Sig[0,2], 0.0)
EQ(Sig[1,0], 0.0)
EQ(Sig[1,2], 0.0)
EQ(Sig[2,0], 0.0)
EQ(Sig[2,1], 0.0)


# Matrix


nelem = 2
nip = 2

mat = GMat.Matrix(nelem, nip)

# all rows: elastic
I = np.ones([nelem, nip], dtype='int')
mat.setElastic(I, K, G)

f = np.zeros((nelem, nip, 3, 3))
for i in range(3):
    for j in range(3):
        f[:, :, i, j] = F[i, j]

Sig = mat.Stress(f)

for e in range(nelem):
    for q in range(nip):
        EQ(Sig[e,q,0,0], G * +2.0 * np.log(1.0 + gamma))
        EQ(Sig[e,q,1,1], G * -2.0 * np.log(1.0 + gamma))

ALLEQ(Sig[:,:,0,1], 0.0)
ALLEQ(Sig[:,:,1,0], 0.0)
ALLEQ(Sig[:,:,:,2], 0.0)
ALLEQ(Sig[:,:,2,:], 0.0)


print('All checks passed')
